path: do not recurse into hidden/dot directories

lucab opened the pull request:
Cargo recursively looks for TOML manifest into child directories,
sometimes getting into unrelated files when exploring hidden
directories (eg. quilt .pc).

This commit lets cargo skip dot directories, to partially avoid
the issues described in #1423.

Signed-off-by: Luca Bruno <lucab@debian.org>

Closes #978

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @wycats (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

alexcrichton commented on src/cargo/ops/cargo_read_manifest.rs:
```diff
@@ -43,8 +43,9 @@ pub fn read_packages(path: &Path, source_id: &SourceId, config: &Config)
-        if dir.file_name().and_then(|s| s.to_str()) == Some(".git") {
+        // Don't recurse into hidden/dot directories
+        if dir.file_name().and_then(|f| f.to_str().and_then(
+                                        |s| Some(s.starts_with(".")))) == Some(true) {
```
Could you tweak the formatting here to perhaps look like:

```rust
let name = dir.file_name().and_then(|s| s.to_str());
if name.map(|s| s.starts_with(".")) == Some(true) {
    return Ok(false)
}
```

alexcrichton commented on src/cargo/sources/path.rs:
```diff
@@ -255,10 +255,13 @@ impl<'cfg> PathSource<'cfg> {
+            match (is_root, name, is_hidden) {
+                (_, _, Some(true)) |
+                (true, Some("target"), Some(false)) |
+                (true, Some("Cargo.lock"), Some(false)) => continue,
```
This may actually be getting to the point where it'll look better broken up:

```rust
let name = dir.file_name().and_then(|s| s.to_str());

// skip dotfile directories
if name.map(|s| s.starts_with(".")) == Some(true) {
    continue
} else if is_root {
    // skip cargo artifacts
    match name {
        "target" | "Cargo.lock" => continue,
        _ => {}
    }
}
```

alexcrichton commented:
Thanks! Could you also add a test for this?

lucab commented:
Rebased according to your comments and added a test. As I didn't find an existing test covering that portion of code, I added a new one in a somehow arbitrary place. Let me know if I missed an existing one, or if it should be moved somewhere else.

alexcrichton commented:
@bors: r+ 11144645fa69cdc5cd6e28c2e8fbff62efc7974a

Thanks!

bors commented:
:hourglass: Testing commit 1114464 with merge e93e27b...

lucab commented:
Digging through the archives, I just discovered that this PR also closes #978.

bors commented:
:sunny: Test successful - [cargo-linux-32](http://buildbot.rust-lang.org/builders/cargo-linux-32/builds/1290), [cargo-linux-64](http://buildbot.rust-lang.org/builders/cargo-linux-64/builds/1280), [cargo-mac-32](http://buildbot.rust-lang.org/builders/cargo-mac-32/builds/1287), [cargo-mac-64](http://buildbot.rust-lang.org/builders/cargo-mac-64/builds/1284), [cargo-win-gnu-32](http://buildbot.rust-lang.org/builders/cargo-win-gnu-32/builds/163), [cargo-win-gnu-64](http://buildbot.rust-lang.org/builders/cargo-win-gnu-64/builds/164), [cargo-win-msvc-32](http://buildbot.rust-lang.org/builders/cargo-win-msvc-32/builds/82), [cargo-win-msvc-64](http://buildbot.rust-lang.org/builders/cargo-win-msvc-64/builds/109)
